libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,171 @@
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+        enum { kSizeType = sizeof(type) };
+
+        struct Node {
+            uint8_t data[kSizeType];
+            bool active;
+        } data_[Count];
+        handle last_;
+
+        inline void setActive_(Node& node, bool active) {
+            node.active = active;
+        }
+
+        inline bool isActive_(Node const& node) const {
+            return node.active;
+        }
+
+        inline type& getItem_(Node& node) {
+            assert(isActive_(node));
+            return *(type*)node.data;
+        }
+
+        inline type const& getItem_(Node const& node) const {
+            assert(isActive_(node));
+            return *(type const*)node.data;
+        }
+
+        inline void destruct_(Node& node) {
+            assert(isActive_(node));
+            auto& item = getItem_(node);
+
+            setActive_(node, false);
+            item.~type();
+        }
+
+        template<typename...__args> inline void construct_(Node& node, __args&&... args) {
+            assert(!isActive_(node));
+
+            new(node.data) type(std::forward<__args>(args)...);
+            setActive_(node, true);
+        }
+
+        inline Node& toNode_(handle h) {
+            assert(h < Count);
+
+            auto& node = *(Node*)&data_[h];
+            return node;
+        }
+
+        inline Node const& toNode_(handle h) const{
+            assert(h < Count);
+
+            auto const& node = *(Node const*)&data_[h];
+            return node;
+        }
+
+        inline void store_(handle h) {
+            last_ = h;
+        }
+
+        inline handle load_() {
+            return last_;
+        }
+
+        inline handle iterateNext_(handle h) {
+            return ++h % Count;
+        }
+
+        inline handle iteratePrev_(handle h) {
+            return (h + (Count - 1)) % Count;
+        }
+
+        inline bool empty_() {
+            return !isActive_(toNode_(last_));
+        }
+
+    public:
+        RingBuffer() : last_(0) {
+            memset(data_, 0, sizeof(data_));
+        }
+
+        ~RingBuffer() {
+            while (!empty_())
+                pop();
+        }
+
+        template<typename...__args>
+        void push(callback&& on_add, callback&& on_remove, __args&&... args) {
+            auto const h = iterateNext_(last_);
+            auto& node = toNode_(h);
+
+            if (isActive_(node)) {
+                if (nullptr != on_remove) {
+                    on_remove(h, getItem_(node));
+                }
+                destruct_(node);
+            }
+
+            construct_(node, std::forward<__args>(args)...);
+            store_(h);
+
+            if (nullptr != on_add) {
+                on_add(h, getItem_(node));
+            }
+        }
+
+        void pop() {
+            auto& node = toNode_(last_);
+            if (isActive_(node)) {
+                store_(iteratePrev_(last_));
```

Comment:
the name `store_` does not correspond to the behaviour. could be something like `setLastUsed`